doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+
+Resolves: https://pagure.io/freeipa/issue/4510
+
+How Google does it: https://support.google.com/accounts/answer/185833?hl=en
+
+## Use cases
+
+### Use with password-only authentication
+
+The user can set multiple passwords per account. Each password can be then used
+for authenticating on only one device. Various passwords can as well be used to
+authenticate to various services or their sets. These divisions can even be
+combined, thus - in the most extreme case - having one password for
+authenticating to one specific service on one particular device. Granularity of
+such divisions will be fully up to user's discretion.
+
+### Use as a part of 2FA (OTP)
+
+When 2FA will be enabled, it will work (and be required) only for the user's
+primary password, not for any of the app passwords, as that would cause too much
+trouble.
+
+### Use as a fallback mechanism when 2FA is activated, but the client app does not support it
+
+As 2FA will not be applied to app passwords, the user can use an app password to
+log into a service that does not support 2FA.
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+App passwords will be stored in a new LDAP subtree in the form
+```
+dn: uid=user-1,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-1
+displayName: My eMail app password for tablet
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
+
+dn: uid=user-2,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-2
+displayName: My eMail app password for smartphone
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
+```
+Every app will have its own subtree in which users will be able to create their
+objects if they will want to use that app - this means objects containing app
+passwords will not be grouped by users, but by apps. The object containing an
+app password will have the following attributes:
+* `uid` - will identify the object, and will also contain a number - this allows
+  for one user having multiple passwords for the same app, its value will be
+  enforced by LDAP ACIs
+* `displayname` - the descriptive label of the password, search display
+  attribute
+* `appname` - the name of the app the password is used for, it is the same as
+  the CN of the parent entry of the object
+* `entryUUID` - a UUID, which will be useful to uniquely identify a password
+  even in extreme cases when `uid`, `displayname` and `appname` will be the same
+  for some password; `entryUUID` will be also used to identify the app password
+  to delete
+* `userpassword` - will hold the app password itself for LDAP
+* `random` - auxiliary virtual attribute, a flag denoting whether to generate
+  the password
+* `randompassword` - auxiliary virtual attribute, stores the generated password
+  for display in the POST callback of the modification command
```

Comment:
So randompassword would not be an attribute in the entry just something return to the user ala host-add foo.example.test --random ?

Reply:
Exactly. This is inspired by such functionality in `user-add`, but I presume that it works analogically in `host-add`.